cubed/array_api/dtypes.py

```diff
@@ -86,3 +87,35 @@
     "complex floating-point": _complex_floating_dtypes,
     "floating-point": _floating_dtypes,
 }
+
+
+# A Cubed-specific utility.
+def _validate_and_define_dtype(x, dtype=None, *, allowed_dtypes=("numeric",), fname=None, device=None):
+    """Ensure the input dtype is allowed. If it's None, provide a good default dtype."""
+    dtypes = __array_namespace_info__().default_dtypes(device=device)
+
+    # Validate.
+    is_invalid = all(x.dtype not in _dtype_categories[a] for a in allowed_dtypes)
+    if is_invalid:
+        errmsg = f"Only {' or '.join(allowed_dtypes)} dtypes are allowed"
+        if fname:
+            errmsg += f" in {fname}"
+        raise TypeError(errmsg)
+
+    # Choose a good default dtype, when None
+    if dtype is None:
+        if x.dtype in _boolean_dtypes:
+            dtype = dtypes["integral"]
+        elif x.dtype in _signed_integer_dtypes:
+            dtype = dtypes["integral"]
+        elif x.dtype in _unsigned_integer_dtypes:
+            # Type arithemetic to produce an unsinged integer dtype at the same default precision.
+            dtype = nxp.dtype(dtypes["integral"].str.replace("i", "u"))
+        elif x.dtype == _complex_floating_dtypes:
+            dtype = dtypes["complex floating"]
+        elif x.dtype == _real_floating_dtypes:
+            dtype = dtypes["real floating"]
```

Comment:
We should remove the two cases for complex and real floating, since `sum` and `prod` should ***not*** upcast in this case (see https://data-apis.org/array-api/latest/API_specification/generated/array_api.sum.html#array_api.sum).

Also, `nansum` should be the same as `sum`. It's not in the spec, but NumPy behaves the same (version 2.1.1):

```python
>>> np.nansum(np.array([1.0, 2.0], dtype=np.float32))
np.float32(3.0)
```